### ISSUE TITLE
sd-process-download failing to process submission

### ISSUE THREAD

**Issue by redshiftzero** [freedomofpress/securedrop-workstation#127]
# Description

@harlo hit some issues with submission processing via `sd-process-download`. The `sd-journalist` AppVM was based on `whonix-ws-14`.

# Steps to Reproduce

1. `make update-fedora-templates`, `make update-whonix-templates`, `make all` on `master`
2. Attempt to download a submission from your staging journalist interface

# Expected Behavior

The submission is processed and decrypted without issue

# Actual Behavior

`sd-process-download` does not process the download and pass the file to the next stage of the submission processing pipeline

# Comments

She ran manually and had the following output:

```
user@host:~$ python /usr/local/bin/sd-process-download /tmp/mozilla_user0/impermeable_aftermath--2018-08-23--19-25-21.zip
Traceback (most recent call last):
  File "/usr/local/bin/sd-process-display", line 2, in <module>
    import PyQt4.QtCore as QtCore
ImportError: No module named PyQt4.QtCore
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
/tmp/tmpe7X3dl is still not a file
^CTraceback (most recent call last):
  File "/usr/local/bin/sd-process-download", line 58, in <module>
    wait_for_sigfile(sigfile)
  File "/usr/local/bin/sd-process-download", line 49, in wait_for_sigfile
    time.sleep(0.5)
KeyboardInterrupt
```

**Comment by conorsch** [comment]
Running `sudo apt install -y pyqt4-dev-tools` inside `sd-journalist` appears to resolve. We'll have to get that logic into the Salt config.

**Comment by conorsch** [comment]
Confirmed fix locally via salt patches. Will have a PR up shortly, along with test updates.

**Comment by conorsch** [comment]
Correction, it's `python-qt4` that should be installed via apt. Having trouble getting the salt commands to trigger. The [docs are straightforward](https://www.qubes-os.org/doc/salt/), but the changes in branch `127-fix-sd-journalist-feedback` don't actually trigger installation of the package in `sd-journalist`. Tinkering with qubesctl commands to resolve...

**Comment by kushaldas** [comment]
In the latest master, `make test` fails for me.

```
$ make test
python -m unittest discover tests
EF
.
.
.
.
./bin/cat: /usr/local/bin/do-not-open-here: No such file or directory
E/bin/cat: /usr/local/bin/move-to-svs: No such file or directory
E/bin/cat: /usr/local/bin/sd-process-display: No such file or directory
E/bin/cat: /usr/local/bin/sd-process-download: No such file or directory
E/bin/cat: /usr/local/bin/sd-process-feedback: No such file or directory
E...........
.
.
.
.
======================================================================
ERROR: test_dispvm (unittest.loader.ModuleImportFailure)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_dispvm
Traceback (most recent call last):
File "/usr/lib64/python2.7/unittest/loader.py", line 254, in _find_tests
module = self._get_module_from_name(name)
File "/usr/lib64/python2.7/unittest/loader.py", line 232, in _get_module_from_name
__import__(name)
File "/home/kdas/securedrop-workstation/tests/test_dispvm.py", line 9, in <module>
import qubes.tests
ImportError: No module named qubes.tests


======================================================================
ERROR: test_do_not_open_here (test_journalist_vm.SD_Journalist_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
File "/home/kdas/securedrop-workstation/tests/test_journalist_vm.py", line 21, in test_do_not_open_here
"sd-journalist/do-not-open-here")
File "/home/kdas/securedrop-workstation/tests/base.py", line 55, in assertFilesMatch
remote_content = self._get_file_contents(remote_path)
File "/home/kdas/securedrop-workstation/tests/base.py", line 51, in _get_file_contents
"/bin/cat {}".format(path)])
File "/usr/lib64/python2.7/subprocess.py", line 219, in check_output
raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command '['qvm-run', '-p', 'sd-journalist', '/bin/cat /usr/local/bin/do-not-open-here']' returned non-zero exit status 1

======================================================================
ERROR: test_move_to_svs (test_journalist_vm.SD_Journalist_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
File "/home/kdas/securedrop-workstation/tests/test_journalist_vm.py", line 13, in test_move_to_svs
"sd-journalist/move-to-svs")
File "/home/kdas/securedrop-workstation/tests/base.py", line 55, in assertFilesMatch
remote_content = self._get_file_contents(remote_path)
File "/home/kdas/securedrop-workstation/tests/base.py", line 51, in _get_file_contents
"/bin/cat {}".format(path)])
File "/usr/lib64/python2.7/subprocess.py", line 219, in check_output
raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command '['qvm-run', '-p', 'sd-journalist', '/bin/cat /usr/local/bin/move-to-svs']' returned non-zero exit status 1

======================================================================
ERROR: test_sd_process_display (test_journalist_vm.SD_Journalist_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
File "/home/kdas/securedrop-workstation/tests/test_journalist_vm.py", line 29, in test_sd_process_display
"sd-journalist/sd-process-display")
File "/home/kdas/securedrop-workstation/tests/base.py", line 55, in assertFilesMatch
remote_content = self._get_file_contents(remote_path)
File "/home/kdas/securedrop-workstation/tests/base.py", line 51, in _get_file_contents
"/bin/cat {}".format(path)])
File "/usr/lib64/python2.7/subprocess.py", line 219, in check_output
raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command '['qvm-run', '-p', 'sd-journalist', '/bin/cat /usr/local/bin/sd-process-display']' returned non-zero exit status 1

======================================================================
ERROR: test_sd_process_download (test_journalist_vm.SD_Journalist_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
File "/home/kdas/securedrop-workstation/tests/test_journalist_vm.py", line 17, in test_sd_process_download
"sd-journalist/sd-process-download")
File "/home/kdas/securedrop-workstation/tests/base.py", line 55, in assertFilesMatch
remote_content = self._get_file_contents(remote_path)
File "/home/kdas/securedrop-workstation/tests/base.py", line 51, in _get_file_contents
"/bin/cat {}".format(path)])
File "/usr/lib64/python2.7/subprocess.py", line 219, in check_output
raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command '['qvm-run', '-p', 'sd-journalist', '/bin/cat /usr/local/bin/sd-process-download']' returned non-zero exit status 1

======================================================================
ERROR: test_sd_process_feedback (test_journalist_vm.SD_Journalist_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
File "/home/kdas/securedrop-workstation/tests/test_journalist_vm.py", line 25, in test_sd_process_feedback
"sd-journalist/sd-process-feedback")
File "/home/kdas/securedrop-workstation/tests/base.py", line 55, in assertFilesMatch
remote_content = self._get_file_contents(remote_path)
File "/home/kdas/securedrop-workstation/tests/base.py", line 51, in _get_file_contents
"/bin/cat {}".format(path)])
File "/usr/lib64/python2.7/subprocess.py", line 219, in check_output
raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command '['qvm-run', '-p', 'sd-journalist', '/bin/cat /usr/local/bin/sd-process-feedback']' returned non-zero exit status 1

======================================================================
FAIL: test_accept_sd_xfer_extracted_file (test_sd_whonix.SD_Whonix_Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
File "/home/kdas/securedrop-workstation/tests/test_sd_whonix.py", line 20, in test_accept_sd_xfer_extracted_file
self.assertFileHasLine("/usr/local/etc/torrc.d/50_user.conf", line)
File "/home/kdas/securedrop-workstation/tests/base.py", line 72, in assertFileHasLine
raise AssertionError(msg)
AssertionError: File /usr/local/etc/torrc.d/50_user.conf does not contain expected line HidServAuth avgfxawdn6c3coe3.onion Il8Xas7uf6rjtc0LxYwhrx

----------------------------------------------------------------------
Ran 27 tests in 78.106s

FAILED (failures=1, errors=6)
Makefile:100: recipe for target 'test' failed
make: *** [test] Error 1



```

**Comment by conorsch** [comment]
The failing test results that @kushaldas reports above are the same as what I see locally on latest master. I'll note also that during salt provisioning runs against e.g. `sd-journalist`, a temporary management VM spins up, i.e. `disp-mgmt-sd-journalist`. That _should_ only happen when the target vm is a TemplateVM, but sd-journalist VM is an AppVM. (See #131 for using TemplateVMs across the board.)

@joshuathayer has not yet been able to reproduce this failure, so I suspect there's a bit of dev env drift somewhere that we need to investigate.

**Comment by eloquence** [comment]
I've split the failure Kushal reported into a separate issue for now, which needs to be investigated at a higher priority. The failure Harlo experienced may or may not be related, and we can debug further once we have unbroken master.

**Comment by conorsch** [comment]
We have a new blocker on the feedback pipeline, caused by merge of the Python3 conversion in #144. Hat tip to @emkll for identifying. Example exception:

```
user@host:~$ python3 /usr/local/bin/sd-process-download /tmp/mozilla_user0/all--2018-09-13--19-18-30.zip
/tmp/tmpjx5ke6cp is still not a file
Traceback (most recent call last):
  File "/usr/local/bin/sd-process-download", line 84, in <module>
    c.communicate(input="DOWNLOAD_BUNDLE_CREATED\n")
  File "/usr/lib/python3.5/subprocess.py", line 786, in communicate
    self._stdin_write(input)
  File "/usr/lib/python3.5/subprocess.py", line 741, in _stdin_write
    self.stdin.write(input)
TypeError: a bytes-like object is required, not 'str'
user@host:~$ Traceback (most recent call last):
  File "/usr/local/bin/sd-process-feedback", line 12, in <module>
    os.write(fifo, arg)
TypeError: a bytes-like object is required, not 'str'
```

Bytes and strings, the joy of Python3 conversions! Should be straightforward to resolve, will open a WIP PR to collaborate with @emkll.

**Comment by redshiftzero** [comment]
resolved by #149
